### PR TITLE
Constraint checks

### PR DESCRIPTION
**Pull request by jchavarri**
This PR adds:
- some initial checking of installed npm packages versus the constraints found on opam depexts file that include the `npm-version` variable
- some basic error handling
- tests for the above

I think this is enough to release a version to opam, the only remaining thing would be to lift the OCaml compiler constraint so that the opam plugin also works with 5.0.

### CODE REVIEW

File: src/main.ml [hunk 1]
```diff
@@ -57,32 +84,69 @@ let check_npm_deps cli =
...
+                      in
+                      let installed_version =
+                        try Ok (RunAsync.runExn (Of_package_json.read path))
+                        with _exn -> Error ()
+                      in
+                      match installed_version with
+                      | Error () ->
+                          print_endline
+                            (Printf.sprintf
+                               "Error: opam package \"%s\" requires npm package \"%s\" \
```
Comment (jchavarri) [inline src/main.ml:119]: I wonder if the command should return an error code when some error has been found?

File: src/main.ml [hunk 2]
```diff
@@ -10,11 +10,24 @@ let match_version_against_constraint version formula =
   let pv = Version.parseExn version in
   Formula.DNF.matches ~version:pv pf
 
+module Of_package_json = struct
+  open EsyLib
```
Comment (davesnx) [inline src/main.ml:14]: an opam plugin using Esy code... the revolution here it comes

File: src/main.ml (hunk 1 above)
Comment (davesnx) [reply src/main.ml:119]: I can see a flag with an option, but for now it's perfect to just print it IMO

Comment (jchavarri) [reply src/main.ml:119]: I meant the opam plugin binary status code returned after the process ends.

Comment (jchavarri) [reply src/main.ml:119]: E.g. if no errors are found, we'd return 0, but if some error is found, we can return 1.

Comment (jchavarri) [reply src/main.ml]: Added a flag `--dry-run` in https://github.com/jchavarri/opam-check-npm-deps/pull/2/commits/c14926612423f0b6e53b69c3b983e1355ae07558.

File: test/main.t [hunk 3]
```diff
@@ -1,3 +1,154 @@
...
+  >     "react": "^18.2.0",
+  >     "react-dom": "^18.2.0"
+  >   }
+  > }
+  > EOF
+
+  $ yarn install --silent
+
+  $ opam-check-npm-deps
+  Error: opam package "test.dev" requires npm package "react-dom" with constraint "^16.0.2", but the version installed is "18.2.0"
```
[the remark this answers is not in the document]
Comment (jchavarri) [reply test/main.t]: Good point, fixed in https://github.com/jchavarri/opam-check-npm-deps/pull/2/commits/cfe4d54f5215fcd8c13807503c2ad028a9cc7bd6.

**Comment by jchavarri** [comment]
One remaining thing I'm not that happy about is that the plugin will check for `node_modules` from the folder where it's called from:

https://github.com/jchavarri/opam-check-npm-deps/blob/cfe4d54f5215fcd8c13807503c2ad028a9cc7bd6/src/main.ml#L113

Maybe this is fine for now, and we can think about smarter ways to find the `node_modules` folder in the future.

**Comment by davesnx** [comment]
Sounds good for a 0.0.1. Possible ways to improve:
- Adding a `--root` flag 
- A mechanism for looking into git folder
- Recursively going up for a node_modules directory

File: src/main.ml [hunk 4]
```diff
@@ -3,18 +3,32 @@ open Cmdliner
 (* Inspired by https://gitlab.ocamlpro.com/louis/opam-custom-install *)
 
 let check_npm_deps_doc = "Check for npm depexts inside the node_modules folder"
+let error_found = ref false
```
Comment (anmonteiro) [inline src/main.ml:6]: should this be moved inside the function that uses it?

File: src/main.ml [hunk 5]
```diff
@@ -3,18 +3,32 @@ open Cmdliner
...
   let pv = Version.parseExn version in
   Formula.DNF.matches ~version:pv pf
 
+module Of_package_json = struct
+  open EsyLib
+
+  type t = { version : string [@default "0.0.0"] }
+  [@@deriving of_yojson { strict = false }]
+
+  let read path =
```
Comment (anmonteiro) [inline src/main.ml:20]: is this `read_version`?

File: src/main.ml [hunk 6]
```diff
@@ -24,10 +38,22 @@ let depexts nv opams =
...
-      OpamSysPkg.Set.empty
+        match has_npm with
+        | false -> npm_pkgs_and_constraints
+        | true -> (
+            match filter with
+            | OpamTypes.FOp (_a, `Eq, FString npm_constraint) ->
+                (names, npm_constraint) :: npm_pkgs_and_constraints
+            | _ ->
+                Printf.printf
+                  "Warning: package %s includes an invalid npm-version \
```
Comment (anmonteiro) [inline src/main.ml:49]: This message explains what's wrong but not how to fix it.

**Comment by jchavarri** [comment]
@davesnx @anmonteiro thanks for the reviews. I think I tackled all the comments, I will merge, then work on testing OCaml 5, and then cut a first release to submit to opam.

A couple of issues that originated from this PR have been created:

- https://github.com/jchavarri/opam-check-npm-deps/issues/5
- https://github.com/jchavarri/opam-check-npm-deps/issues/4

File: src/main.ml (hunk 6 above)
Comment (jchavarri) [reply src/main.ml:48]: Added some suggestion to fix the issue in https://github.com/jchavarri/opam-check-npm-deps/pull/2/commits/e8d5c111ee0fc1fabda178429ecadad110586c54.

**Pull request merged**